pkg/arvo/app/dojo.hoon

```diff
@@ -1349,16 +1350,33 @@
   !>(state)
 ::
 ++  on-load
-  |=  =old-state=vase
-  =/  old-state  !<(house old-state-vase)
-  `..on-init(state old-state)
+  |=  old=vase
+  ?:  ?=(%6 +<.old)
+    `..on-init(state !<(house old))
+  =/  old-5  !<([%5 egg=@u hoc=(map id session)] old)
+  =/  =house  [%6 egg.old-5 hoc.old-5 *(set ship)]
+  `..on-init(state house)
 ::
 ++  on-poke
   |=  [=mark =vase]
   ^-  (quip card:agent:gall _..on-init)
   =^  moves  state
     ^-  (quip card:agent:gall house)
     ?+  mark  ~|([%dojo-poke-bad-mark mark] !!)
+    ::
+        %noun
+      ?+    q.vase  ~|([%dojo-poke-bad-noun vase] !!)
+          [%grant-access @p]
+        `state(acl (~(put in acl) +.q.vase))
+      ::
+          [%revoke-access @p]
+        `state(acl (~(del in acl) +.q.vase))
```

Comment:
I believe this also needs to kick any active subscriptions; currently, once you `|link`, you remain linked even after your access is revoked. Is that as simple as returning a `%kick` card here?

Reply:
Yes, this is what the `(unit ship)` in `%kick` is used for.  You don't have to specify the path.

Reply:
This is great! It'd be nice to add generators (and probably marks) for these administrative pokes, then people could `:dojo|grant ~zod` or `:dojo|revoke ~zod`. You just put the generators in `gen/dojo/` to enable match that syntax.

Reply:
Done. `%kick` still doesn't seem to be working, though; `~zod` remains linked to `~nec` even after being kicked. Not sure what's going on.

Reply:
Hmm, it looks like it actually only kicks the subscriptions you specify, and the ship is just a filter on top of that.  I guess just specify the subscription path.  I think at some point we should change gall to have the semantics I thought it had.

https://github.com/urbit/urbit/blob/69b6495d3d4c8cd3d81c5842a2cc90bf8059d63c/pkg/arvo/sys/vane/gall.hoon#L1108-L1118

Reply:
Great, it's working now -- zod gets unlinked as soon as nec runs `:dojo|revoke ~zod`.

Unfortunately this means that we now have `(cat 3 'drum_' (scot %p ship))` floating around in *four* separate places, two of those in Gall apps. I'd really like to consolidate those if possible.

Reply:
uhh, I'd just kick all paths for that ship - as stands I think the acl also lets them connect _web_ dojo which then doesn't get kicked
(or any other hypothetical future client)

Reply:
This is true.  This is really a bug in gall.  There's code to clear all subscriptions from a ship if list of paths is null, but it's inside that `+turn`, which means it doesn't ever get run.  We should handle the `?~  target-path` case in `+ap-ducts-from-paths` before the `+zing` instead of in `+ap-ducts-from-path` (note pluralization).  In fact, `+ap-ducts-from-path` shouldn't take a `(unit path)` at all, but a `path`.

@lukechampine I can get to this in the next few days, or you're welcome to do it.  Just test it by taking away the subscription path in your `%kick`.

Reply:
I think I got it! Implementation can probably be cleaned up a bit, but it appears to work.